filecoin-proofs/src/api/seal.rs

```diff
@@ -538,6 +541,345 @@ pub fn seal_commit_phase2<Tree: 'static + MerkleTreeTrait>(
+                .expect("failed to construct multi proof from bytes")
+                .circuit_proofs,
+            );
+
+            acc
+        });
+    trace!(
+        "aggregate_seal_commit_proofs called with {} commit_outputs",
+        commit_outputs.len()
+    );
+
+    let target = if commit_outputs.len() == 1 {
+        2
+    } else {
+        commit_outputs.len().next_power_of_two()
+    };
+    trace!("aggregate_seal_commit_proofs will pad to target {}", target);
+    ensure!(target > 1, "cannot aggregate less than two proofs");
+
+    // If we're not at the pow2 target, duplicate the last element until we are.
+    while proofs.len() != target {
+        ensure!(
+            proofs.last().is_some(),
+            "invalid last proof for duplication"
+        );
+        proofs.push(
+            proofs
+                .last()
+                .expect("failed to access last proof for duplication")
+                .clone(),
+        );
+    }
+    trace!(
+        "padded proofs from {} to {}",
+        commit_outputs.len(),
+        proofs.len()
+    );
+
+    ensure!(
+        proofs.len().next_power_of_two() == proofs.len(),
+        "proof count must be a power of 2 for aggregation"
+    );
+    ensure!(
+        proofs.len() <= SRS_MAX_PROOFS_TO_AGGREGATE,
+        "proof count for aggregation is larger than the max supported value"
+    );
+
+    let srs_keys = get_stacked_srs_key::<Tree>(porep_config, proofs.len())?;
+    let aggregate_proof = StackedCompound::<Tree, DefaultPieceHasher>::aggregate_proofs(
+        &srs_keys.prover_srs,
+        proofs.as_slice(),
+    )?;
+    let aggregate_proof_bytes = serialize(&aggregate_proof)?;
+
+    info!("aggregate_seal_commit_proofs:finish");
+
+    Ok(aggregate_proof_bytes)
+}
+
+// 'aggregate_proof_bytes' is the serialized aggregate proof returned from 'aggregate_seal_commit_proofs'.
+// 'commit_inputs' is an ordered list of inputs returned from 'seal_commit_phase2_for_aggregation'.  The
+// order of that list must match the order of the 'commit_outputs' provided to 'aggregate_seal_commit_proofs'
+// in order for verification to work properly.
+//
+// aggregated_proofs_len is not the same as commit_inputs.len, since commit_inputs is a combined list and
+// aggregated_proofs_len describes the count that went into generating the aggregate proof initially.
+pub fn verify_aggregate_seal_commit_proofs<Tree: 'static + MerkleTreeTrait>(
+    porep_config: PoRepConfig,
+    aggregated_proofs_len: usize,
+    aggregate_proof_bytes: AggregateSnarkProof,
+    commit_inputs: Vec<Vec<Fr>>,
+) -> Result<bool> {
+    info!("verify_aggregate_seal_commit_proofs:start");
+
+    ensure!(!commit_inputs.is_empty(), "cannot verify with empty inputs");
+    ensure!(
+        commit_inputs.len() % aggregated_proofs_len == 0,
+        "invalid number of inputs provided"
+    );
+
+    let num_inputs = commit_inputs.len();
+    let num_inputs_per_proof = num_inputs / aggregated_proofs_len;
```

Comment:
isn't this supposed to be fixed ? i.e. shouldn't we read that info from .. porep config or stg ? 
What we want to do is twofold:
1. ensure there are always X public inputs **per proof** -> this is different than checking there are X * Y public inputs (where Y is the number of proofs) in total
2. ensure the number of proofs is a power of two, if it's not, we duplicate X public inputs as much as necessary

It feels 1. and 2. are mixed here, and 1. is checking the total number of public inputs rather than the number of public inputs per proofs.